pkg/csi_driver/multishare_controller.go

```diff
@@ -218,6 +220,101 @@ func (m *MultishareController) CreateVolume(ctx context.Context, req *csi.Create
 	return resp, file.StatusError(err)
 }
 
+func (m *MultishareController) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshotRequest) (*csi.CreateSnapshotResponse, error) {
+	klog.Infof("CreateSnapshot called for multishare with request %+v", req)
+	name := req.GetName()
+	volumeID := req.GetSourceVolumeId()
+
+	if acquired := m.volumeLocks.TryAcquire(volumeID); !acquired {
+		return nil, status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, volumeID)
+	}
+	defer m.volumeLocks.Release(volumeID)
+
+	if req.GetParameters() != nil {
+		if _, err := util.IsSnapshotTypeSupported(req.GetParameters()); err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+	}
+	_, location, instanceName, shareName, err := parseSourceVolId(volumeID)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	project := m.cloud.Project
+
+	backupLocation := util.GetBackupLocation(req.GetParameters()) //Optional provided locaiton for cross-region backups
+	backupURI, backupRegion, err := file.CreateBackupURI(location, project, name, backupLocation)
+	if err != nil {
+		klog.Errorf("Failed to create backup URI from given name %s and location %s, error: %v", req.Name, backupLocation, err.Error())
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	existingBackup, err := m.cloud.File.GetBackup(ctx, backupURI)
+	backupExists, err := file.CheckBackupExists(existingBackup, err)
+	if err != nil {
+		return nil, file.StatusError(err)
+	}
+
+	if backupExists {
+		// process existing backup
+
+		snapshot, err := file.ProcessExistingBackup(ctx, existingBackup, volumeID, modeMultishare)
+		if err != nil {
+			return nil, err
+		}
+		return &csi.CreateSnapshotResponse{
+			Snapshot: snapshot,
+		}, nil
+	} else {
+		//no existing Snapshot
+		backupInfo := &file.BackupInfo{
+			Name:               name,
+			SourceVolumeId:     volumeID,
+			Project:            project,
+			Location:           backupRegion,
+			SourceShare:        shareName,
+			SourceInstanceName: instanceName,
+			BackupURI:          backupURI,
+		}
+
+		labels, err := extractBackupLabels(req.GetParameters(), m.driver.config.Name, req.Name)
+		if err != nil {
+			return nil, err
+		}
+		backupInfo.Labels = labels
+		snapshot, err := m.createNewBackup(ctx, backupInfo)
+		if err != nil {
+			return nil, err
+		}
+
+		resp := &csi.CreateSnapshotResponse{
+			Snapshot: snapshot,
+		}
+		return resp, nil
+	}
+}
+
+func (m *MultishareController) createNewBackup(ctx context.Context, backupInfo *file.BackupInfo) (*csi.Snapshot, error) {
+
+	backupObj, err := m.cloud.File.CreateBackup(ctx, backupInfo)
+	if err != nil {
+		klog.Errorf("Create snapshot for volume Id %s failed: %v", backupInfo.SourceVolumeId, err.Error())
```

Comment:
What would happen if the backup creation took some time and timed out? what is the error code returned here? is it abort or internal or something else? concern is that backup creation could push up the SLO signal

Reply:
That's a good question, and I'm not sure. We didn't have any special case handling for singleshare backups, so I didn't put any here. I'll investigate.

Reply:
Discussed with the FS team. They aren't expecting timeouts from this, since the SLO is on the order of hours/days depending on size of volume.